make-subcanon-maps.c

```diff
@@ -0,0 +1,218 @@
+// that is big enough for the 256 million permutations from non-canonicals to canonicals for k=8, even if k<8.
+// So we're allocating 256MBx3=768MB even if we need much less.  I figure anything less than 1GB isn't a big deal
+// these days. It needs to be aligned to a page boundary since we're going to mmap the binary file into this array.
+static kperm Permutations[maxBk] __attribute__ ((aligned (8192)));
+// Here's the actual mapping from non-canonical to canonical, same argument as above wasting memory, and also mmap'd.
+// So here we are allocating 256MB x sizeof(short int) = 512MB.
+// Grand total statically allocated memory is exactly 1.25GB.
+static short int K[maxBk] __attribute__ ((aligned (8192)));
+
+//Needed for BuildGraph
+static TINY_GRAPH *G;
+static int k;
+
+#define MMAP 1 
+// Try to mmap, and if it fails, just slurp in the file (sigh, Windoze)
+void *Mmap(void *p, size_t n, int fd)
+{
+#if MMAP
+    void *newPointer = mmap(p, n, PROT_READ, MAP_PRIVATE|MAP_FIXED, fd, 0);
+    if(newPointer == MAP_FAILED)
+#endif
+    {
+#if !__WIN32__ && !__CYGWIN__ // it will always fail on Windoze so don't bother with a warning
+	Warning("mmap failed");
+#endif
+	if(read(fd, p, n) != n)
+	    Fatal("cannot mmap, or cannot read the file, or both");
+    }
+    return p;
+}
+
+static void BuildGraph(int Gint)
+{
+    int i, j, bitPos=0;
+    int Gint2 = Gint;  // Gint2 has bits nuked as they're used, so when it's zero we can stop.
+    TinyGraphEdgesAllDelete(G);
+#if LOWER_TRIANGLE
+    for(i=k-1;i>0;i--)
+    {
+	for(j=i-1;j>=0;j--)
+#else	// UPPER_TRIANGLE
+    for(i=k-2;i>=0;i--)
+    {
+	for(j=k-1;j>i;j--)
+#endif
+	{
+	    if(!Gint2) break;
+	    int bit = (1 << bitPos);
+	    if(Gint & bit)
+		TinyGraphConnect(G,i,j);
+	    Gint2 &= ~bit;
+	    bitPos++;
+	}
+	if(!Gint2) break;
+    }
+}
+
+#define LOWER_TRIANGLE 1
+// Given a TINY_GRAPH and k, return the integer ID cretaed from one triangle (upper or lower) of the adjacency matrix.
+static int TinyGraph2Int(TINY_GRAPH *g, int numNodes)
+{
+    int i, j, bitPos=0, Gint = 0, bit;
+    
+#if LOWER_TRIANGLE	// Prefer lower triangle to be compatible with Ine Melckenbeeck's Jesse code.
+    for(i=numNodes-1;i>0;i--)
+    {
+        for(j=i-1;j>=0;j--)
+#else   // UPPER_TRIANGLE // this is what we used in the original faye code and paper with Adib Hasan and Po-Chien Chung.
+    for(i=numNodes-2;i>=0;i--)
+    {
+        for(j=numNodes-1;j>i;j--)
+#endif
+        {
+	    if(TinyGraphAreConnected(g,i,j))
+	    {
+		bit = (1 << bitPos);
+		Gint |= bit;
+	    }
+            bitPos++;
+        }
+    }
+    return Gint;
+}
+
+const char *byte_to_binary(int x)
+{
+    static char b[9];
+    b[0] = '\0';
+
+    int z;
+    for (z = 128; z > 0; z >>= 1)
+    {
+        strcat(b, ((x & z) == z) ? "1" : "0");
+    }
+
+    return b;
+}
+
+
+int main(int argc, char* argv[]) {
+    /*
+    read from canon list files

```

Comment:
Probably should say, "read from the canon_list file specified by *k* on the command line..."